File: src/etch/bindings/java/transport/DefaultMessageHandler.java
/* $Id$
 *
 * Created by wert on Mar 26, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.transport;

import metreos.core.Who;
import etch.bindings.java.msg.Message;
import etch.bindings.java.support.StubBase;

/**
 * Description of DefaultMessageHandler.
 * @param <T> The type of the stub
 */
abstract public class DefaultMessageHandler<T extends StubBase>
	implements MessageHandler<MailboxMessageSource>
{
	/**
	 * Constructs the DefaultMessageHandler.
	 */
	public DefaultMessageHandler()
	{
		// nothing to do.
	}
	
	/**
	 * @param src
	 * @return new stub
	 */
	abstract protected T newStub( MailboxMessageSource src );
	
	@Override
	abstract public String toString();

	private T stub;

	public boolean message( Who sender, Message msg ) throws Exception
	{
		return stub.message( sender, msg );
	}
	
	public MailboxMessageSource getSource()
	{
		return src;
	}
	
	public void setSource( MailboxMessageSource src )
	{
		this.src = src;
	}
	
	private MailboxMessageSource src;
	
	public Object sessionQuery( Object query )
	{
		throw new UnsupportedOperationException( "unknown query: "+query );
	}
	
	public void sessionControl( Object control, Object value )
	{
		throw new UnsupportedOperationException( "unknown control: "+control );
	}
	
	public void sessionNotify( Object event )
	{
		if (event.equals( Session.UP ))
		{
			stub = newStub( src );
			return;
		}

		if (event.equals( Session.DOWN ))
		{
			stub = null;
			return;
		}
		
		// ignore.
	}
}

File: src/etch/bindings/csharp/transport/DefaultMessageHandler.cs
using System;
using System.Text;
using System.Collections.Generic;
using Etch.Msg;
using Etch.Support;

namespace Etch.Transport
{
    abstract public class DefaultMessageHandler<T> : MessageHandler<MailboxMessageSource> 
        where T : StubBase
    {

        /// <summary>
        /// Constructs the default message handler
        /// </summary>
        public DefaultMessageHandler()
        {
            // nothing to do.
        }

        abstract protected T NewStub( MailboxMessageSource src );

        public abstract override string ToString();

        private T stub;

        /// <summary>
        /// 
        /// </summary>
        /// <param name="src"></param>
        /// <param name="sender"></param>
        /// <param name="msg"></param>
        /// Exception:
        ///     throws Exception
        public bool Message( Who sender, Message msg )
        {
            return stub.Message(sender, msg);
        }

        public MailboxMessageSource GetSource()
        {
            return _src;
        }

        public void SetSource( MailboxMessageSource src )
        {
            _src = src;
        }

        private MailboxMessageSource _src;

        public Object SessionQuery( Object query )
        {
            throw new Exception( " unknown query " + query );
        }

        public void SessionControl( Object control, Object value )
        {
            throw new Exception( " unknown control " + control );
        }

        public void SessionNotify( Object eventObj )
        {
            if ( eventObj.Equals( SessionConsts.UP ) )
            {
                stub = NewStub( _src );
                return;
            }

            if ( eventObj.Equals( SessionConsts.DOWN ) )
            {
                stub = null;
                return;
            }
        }
    }
}
